File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        std::stack<char> cstk;
        
        for (auto &c:s) {
            if (!cstk.empty() && 
                       ((cstk.top() == '(' && c == ')') || 
                       (cstk.top() == '[' && c == ']') || 
                       (cstk.top() == '{' && c == '}'))
                      ) {
                    cstk.pop();
                }  else
                cstk.push(c);
        }
        return cstk.empty();
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut cstk = Vec::new();
        let cstring:Vec<char> = s.chars().collect();
        for c in cstring {
            if !cstk.is_empty() &&
                    ((cstk.last().unwrap() == &'[' && c == ']') || 
                    (cstk.last().unwrap() == &'{' && c == '}') ||
                    (cstk.last().unwrap() == &'(' && c == ')')) {
                        cstk.pop();
                }  else {
                cstk.push(c);
            }
        }
        cstk.is_empty()
    }
}